client/verta/verta/runtime.py

```diff
@@ -0,0 +1,93 @@
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict
+import json
+import threading
+import os
+
+
+_THREAD = threading.local()
+
+def init_thread_context() -> None:
+    """Initialize our thread-local variable for storing logging context."""
+    if not hasattr(_THREAD, "context"):
+        _THREAD.context = defaultdict()
+
+
+def get_thread_context() -> defaultdict:
+    """
+    Return the current thread-local context or initialize a new one.
+    """
+    init_thread_context()
+    return _THREAD.context
+
+
+def set_thread_context(context: defaultdict) -> defaultdict:
+    """
+    Sets the thread-local context, overwriting any existing values.
+    """
+    # init_thread_context()
+    _THREAD.context = context
+    return _THREAD.context
+
+
+def log(**kwargs) -> None:
+    """
+    Log any number of key-value pairs to the current prediction context.
+    Context is stored in thread-local variables.
+
+    Parameters
+    ----------
+    kwargs : Any
+        Arguments and values to be added as key-value pairs to the current
+        logging context. Values must be serializable to JSON.
+
+    Returns
+    -------
+    None
+    """
+    local_context: defaultdict = get_thread_context().copy()
+    local_context.update(**kwargs)
+    set_thread_context(local_context)
+
+
+class context:
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    Uses thread-local variables to store context in a thread-safe manner.
+
+    Parameters
+    ----------
+    kwargs : Any
+        Key-value pairs to be added to the current logging context.
+    Attributes
+    ----------
+    as_json : str
+        A JSON string representation of all the current saved logging context.
+    """
+    def __init__(self, **kwargs):
+        self.new_context = kwargs
+        self.prior_context = defaultdict()
+
+    def __enter__(self):
+        self.prior_context = get_thread_context()
+        updated_context: defaultdict = self.prior_context.copy()
+        updated_context.update(self.new_context)
+        set_thread_context(updated_context)
+        return self
+
+    def __exit__(self, *args):
+        """ Return thread-local context to prior state when exiting this context. """
+        set_thread_context(self.prior_context)
```

Comment:
I think we're going to encounter issues if `ctx.logs()` is ever called after the context is exited.

Reply:
The advantage of writing our own context manager here is that we can store the final logs in a class variable so they could still be accessed after exiting the context.  The thread-local variable holding the logs would still be wiped on exit, but the context manager object would still have the logs if `ctx.logs()` were called outside the context.

I've just updated the PR with that change, we can revert if we change our minds.

Reply:
That's a fair (and cool) idea!